src/odemis/driver/keysight.py

```diff
@@ -481,56 +481,93 @@ def _set_power(self, p: bool) -> bool:
 
         # Immediately update the internal state, to let the setters know they should apply the settings
         self._is_powered = p
-        if p:
-            # Powered on means a square waveform
-            self.setWaveform(self._channel, "SQU")
-            for c, t in self._tracking.items():
-                other_c = 3 - c
-                self.setTracking(other_c, t)
+        try:
+            if p:
+                self._power_device_on()
+            else:
+                self._power_device_off()
+        except Exception:
+            # If an error occurs, the device might be in some sort of inconsistent state. It's pretty
+            # hard to do "something clever". However, at least we should assume intenal state is synchronized
+            # with the VA value (which will not be updated). This way, try to set the VA again, will
+            # allow to try again doing the same thing.
+            self._is_powered = not p
+            raise
 
-            for c, lim in enumerate(self._limits):
-                self.setVoltageMin(c + 1, lim[0])
-                self.setVoltageMax(c + 1, lim[1])
+        return p
```

Comment:
I wonder why you want to return something in a setter? It's not like the value will be different from the input parameter.

Reply:
This is the behaviour of setters, as defined by VAs. They take the value to be set, and return the "actual" value set. For a boolean, it's usually always the same, but for floats, like the exposure time, it can easily be different. You can find an example here with the period: the setter computes the frequency based on the period, then it's passed to the hardware, which rounds it to which ever frequency is supported, and then this frequency is converted back to the "actual" period accepted by the hardware.